Google.Api.Gax.Grpc/Rest/RestChannel.cs

```diff
@@ -105,51 +117,97 @@ private async Task AddAuthHeadersAsync(HttpRequestMessage request, RestMethod me
+        /// <see cref="Status"/> is not.
+        /// For being able to surface error details in <see cref="Status"/> we need to be
+        /// able to call <see cref="HttpContent.ReadAsStringAsync"/> which is an async method,
+        /// and thus cannot be done, without blocking, on the sync function that obtains the 
+        /// <see cref="Status"/> in the <see cref="AsyncUnaryCall{TResponse}"/>.
+        /// So we need to make async content reading part of sending the call and not part of
+        /// building the TResponse as it is now.
+        /// This class is just a convenient wrapper for passing together the <see cref="HttpResponseMessage"/>
+        /// and it's read response.
+        /// </summary>
+        internal class ReadHttpResponseMessage
+        {
+            private HttpResponseMessage OriginalResponseMessage { get; }
+
+            private readonly string _content;
+            private readonly ExceptionDispatchInfo _readException;
+
+            internal string Content
+            {
+                get
+                {
+                    _readException?.Throw();
+                    return _content;
+                }
+            }
+
+            internal ReadHttpResponseMessage(HttpResponseMessage response, string content) =>
+                (OriginalResponseMessage, _content) = (response, content);
+
+            internal ReadHttpResponseMessage(HttpResponseMessage response, ExceptionDispatchInfo readException) =>
+                (OriginalResponseMessage, _readException) = (response, readException);
+
+            internal Metadata GetHeaders()
+            {
+                // TODO: This could be very wrong. I don't know what headers we should really return, and I don't know about semi-colon joining.
+                var metadata = new Metadata();
+                foreach (var header in OriginalResponseMessage.Headers)
+                {
+                    metadata.Add(header.Key, string.Join(";", header.Value));
+                }
+                return metadata;
+            }
+
+            internal Status GetStatus()
+            {
+                var grpcStatus = OriginalResponseMessage.StatusCode switch
+                {
+                    HttpStatusCode.OK => StatusCode.OK,
+                    HttpStatusCode.Unauthorized => StatusCode.Unauthenticated,
+                    HttpStatusCode.Forbidden => StatusCode.PermissionDenied,
+                    HttpStatusCode.BadRequest => StatusCode.InvalidArgument,
+                    HttpStatusCode.InternalServerError => StatusCode.Internal,
+                    HttpStatusCode.NotFound => StatusCode.NotFound,
+                    HttpStatusCode.Conflict => StatusCode.FailedPrecondition,
+                    // TODO: Others!
+                    _ => StatusCode.Unknown
+                };
+
+                return new Status(grpcStatus,
+                    // Notice that here, if there was an exception reading the content
```

Comment:
Right - so this is an exception from "we can't even get the error message" rather than "the call failed and the error message is present, yes?

Is the error message likely to be in JSON that we should parse?

Reply:
Yes to the first part. This is "something went wrong and the request probably didn't even make it to the backend".

And yes, the error message is likely to be in JSON. My original intention was to do the same as we do in Apiary, which is, we attempt to parse, if we succeed then we set the parsed object, else we set the content as a string on the GoogleApiException.

The problem here is that `Grpc.Core.Status` only has a `StatusCode` and `string Detail` so we cannot set anything more complex there. And the `AsyncUnaryCall` expects `Grpc.Core.Status`.

Of course I could inherit from `Status` and add anything. I could even inherit from `RpcException` to have an exception type that only accepts the more specific `Status` type. Not sure if it is worth it, specially since the JSON will have some standard properties that we can expect, but API teams can add their own that we have no idea of, so, in Apiary, we end up adding those as a string I think.
What do you think?

Reply:
I think what's here is sufficient for a beta. Let's see what it looks like in reality, and find out what other languages are doing.